sdk/src/client/node_api/indexer/query_parameters.rs

```diff
@@ -3,310 +3,240 @@
 
 //! Query parameters for output_id requests
 
-use std::fmt;
-
 use serde::{Deserialize, Serialize};
 
-use crate::{
-    client::{node_api::query_tuples_to_query_string, Error, Result},
-    types::block::{address::Bech32Address, slot::SlotIndex},
-};
+use crate::types::block::{address::Bech32Address, output::TokenId, slot::SlotIndex};
 
 // https://github.com/iotaledger/inx-indexer/tree/develop/pkg/indexer
 
-/// Query parameters for output_id requests.
-#[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct QueryParameters(Vec<QueryParameter>);
-
-impl QueryParameters {
-    /// Creates a hashset from a provided vec of query parameters.
-    #[must_use]
-    pub fn new(query_parameters: impl Into<Vec<QueryParameter>>) -> Self {
-        let mut query_parameters = query_parameters.into();
-        query_parameters.sort_unstable_by_key(QueryParameter::kind);
-        query_parameters.dedup_by_key(|qp| qp.kind());
-
-        Self(query_parameters)
-    }
-
-    /// Creates new empty QueryParameters.
-    pub fn empty() -> Self {
-        Self(Vec::new())
-    }
+pub trait QueryParameter {
+    /// Converts parameters to a single String.
+    fn to_query_string(&self) -> Option<String>;
+    fn replace_cursor(&mut self, cursor: String);
+}
 
-    /// Replaces or inserts an enum variant in the QueryParameters.
-    pub fn replace(&mut self, query_parameter: QueryParameter) {
-        match self
-            .0
-            .binary_search_by_key(&query_parameter.kind(), QueryParameter::kind)
-        {
-            Ok(pos) => self.0[pos] = query_parameter,
-            Err(pos) => self.0.insert(pos, query_parameter),
+macro_rules! impl_query_parameters_methods {
+    ($name:ty) => {
+        impl QueryParameter for $name {
```

Comment:
smart to do this using serde, but I'm pretty sure this doesn't need to be a macro, it can just be the default impl with Self: Serialize` bounds

Reply:
Changed